## 2. Hypothesis Testing and Confidence Intervals
### 2.1 Hypothesis Testing

**Overview**

Hypothesis tests are used to confirm or reject assumptions (hypothesis) based on statistical measures. In applied machine learning, one application of hypothesis tests is to test whether two sets of data are from the same distribution or not (e.g., from the same population) or to test whether a set is normally distributed.<br>

There are two hypothesis:<br>
- **H0:**Test holds and is failed to be rejected at some level of significance
- **H1**: Test does not hold and is rejected at some level of significance

To accept or reject H0, we compare the **p-value** to a significance level $\alpha$ (0.1,0.05,0.01).<br>
- If p-value > $\alpha$: Fail to reject H0
- if p-value <= $\alpha$: Reject H0<br>

The p-value is a measure of how likely the data sample would be observed if H0 were true.<br>

Though, not all tests return a p-value. Some tests return "critical values" and their associated significance levels instead. These are usually non-parametric or distribution-free hypothesis tests.<br>
- If test statistic < critical value: Fail to rejetc H0
- If test statistic >= critical value: Reject H0<br>

The mathematical steps (generalized) are as follows:<br>
1. Formt he hypothesis $H_{0}:\mu = \mu_{0}$ and the alternative $H_{a}:\mu > \mu_{0}$ or $H_{a}:\mu < \mu_{0}$ or $H_{a}:\mu \neq \mu_{0}$.<br>
2. Calculate the test statistic $z$:<br>
$z=\frac{\bar{x}-\mu_{0}}{\frac{\sigma}{\sqrt{n}}}$<br>
3. Determine the p-value. Since due to the Central Limit Theorem under $H_{0}$ $Z=\frac{\bar{X}-\mu_{0}}{\frac{\sigma}{\sqrt{n}}}\sim Norm(0,1)$ the p-value is calculated using normal distribution function. Under $H_{0}$ we should see an event of high probability.


**Errors in Statistical Tests**

Due to the probabilistic nature of hypothesis tests, they may suggest an outcome and be mistaken. For example, if $\alpha=0.05$, it suggests that one time in 20 the null hypothesis would be mistakenly rejected or failed to be reject because of the statistical noise in the data sample.<br>

Given a small p-value, we either reject H0 correctly or, with a small probability, we incorrectly rejected H0 and it was actually true. The latter is an example of a **false positive**.<br>

Alternatively, given a large p-value we fail to reject H0, even though H0 should actually be rejected. This would be called a **false negative**.<br>

- **Type I Error:** Incorrect rejection of a true H0 (false positive)
- **Type II Error:** Incorrect failure of rejection of a false H0 (false negative)<br>

Obviously, you can somewhat control the likelihood of such errors by lowering your significance level (e.g., 5-sigma threshold, often used in physics, which is $3*10^{-7}$).<br>

### 2.2 Examples of Hypothesis Tests
#### 2.2.1 Variable Distribution Type Tests (Gaussian)

This chapter describes three tests and gives basic code in Python to test whether your data is gaussian distributed.<br>

**Shapiro-Wilk Test**

Assumptions:<br>
- Obervations are i.i.d. (independent and identically distributed)<br>

Interpretation:<br>
- H0: Sample has gaussian distribution
- H1: Sample does not have gaussian distribution

Python:

In [1]:
from scipy.stats import shapiro

## Normality test
data = [0.873, 2.817, 0.121, -0.945, -0.055, -1.436, 0.360, -1.478, -1.637, -1.869]
stat, p = shapiro(data)

print(p) # Fail to reject H0 --> Gaussian

0.19340917468070984


**D'Agostino's $K^{2}$ Test**

Assumptions:<br>
- Obervastions are i.i.d.

Interpretation:<br>
- H0: Sample has gaussian distribution
- H1: Sample does not have gaussian distribution

Python:

In [2]:
from scipy.stats import normaltest

## Normality test
data = [0.873, 2.817, 0.121, -0.945, -0.055, -1.436, 0.360, -1.478, -1.637, -1.869]
stat, p = normaltest(data)

print(p) # Fail to reject H0 --> Gaussian

0.18342710340675566


C:\Users\wehrm\AppData\Local\Continuum\anaconda3\lib\site-packages\scipy\stats\stats.py:1394: UserWarning: kurtosistest only valid for n>=20 ... continuing anyway, n=10
  "anyway, n=%i" % int(n))


**Anderson-Darling Test**

Assumptions:<br>
- Observations are i.i.d.

Interpretation:<br>
- H0: Sample has gaussian distribution
- H1: Sample does not have gaussian distribution

Python:

In [4]:
from scipy.stats import anderson

## Normality test
data = [0.873, 2.817, 0.121, -0.945, -0.055, -1.436, 0.360, -1.478, -1.637, -1.869]
result = anderson(data)

## get critical value and significance level
for i in range(len(result.critical_values)):
    sl,cv = result.significance_level[i], result.critical_values[i]
    if result.statistic < cv:
        print("Fail to reject H0 (gaussian) at %.1f%% level" %(sl))
    else:
        print("Reject H0 (not gaussian) at %.1f%% level" %(sl))

Fail to reject H0 (gaussian) at 15.0% level
Fail to reject H0 (gaussian) at 10.0% level
Fail to reject H0 (gaussian) at 5.0% level
Fail to reject H0 (gaussian) at 2.5% level
Fail to reject H0 (gaussian) at 1.0% level


#### 2.2.2 Variable Relationship Tests (Correlation)

This sub-chapter lists statistical tests to check if two samples are related and gives the Python code to do so.

**Pearson's Correlation Coefficient**

Tests whether two samples have a linear relationship.<br>

Assumptions:<br>
- Observations are i.i.d.
- Both samples are normally distributed
- Both samples have the same variance

Interpretation:<br>
- H0: Samples are independent
- H1: Samples are dependent

Python:

In [5]:
from scipy.stats import pearsonr

## Test linear correlation
data1 = [0.873, 2.817, 0.121, -0.945, -0.055, -1.436, 0.360, -1.478, -1.637, -1.869]
data2 = [0.353, 3.517, 0.125, -7.545, -0.555, -1.536, 3.350, -1.578, -3.537, -1.579]
stat, p = pearsonr(data1, data2)

print(p) # Reject H0 at 5% level --> samples are dependent

0.027872969514496193


**Spearman's Rank Correlation**

Tests whether two samples have a monotonic relationship (positive increase in X causes positive increase in Y and vice versa - positive correlation).<br>

Assumptions:<br>
- Observations are i.i.d.
- Observations in each sample can be ranked

Interpretation:<br>
- H0: Samples are independent
- H1: Samples are dependent

Python:

In [6]:
from scipy.stats import spearmanr

## Test monotonic correlation
data1 = [0.873, 2.817, 0.121, -0.945, -0.055, -1.436, 0.360, -1.478, -1.637, -1.869]
data2 = [0.353, 3.517, 0.125, -7.545, -0.555, -1.536, 3.350, -1.578, -3.537, -1.579]
stat, p = spearmanr(data1, data2)

print(p) # Reject H0 --> Monotonic relationship

0.0016368033159867143


**Kendall's Rank Correlation**

Tests whether two samples have a monotonic relationship (see above).<br>

Assumptions:<br>
- Observations are i.i.d.
- Observations in each sample can be ranked

Interpretation:<br>
- H0: Samples are independent
- H1: Samples are dependent

Python:

In [7]:
from scipy.stats import kendalltau

## Test monotonic correlation
data1 = [0.873, 2.817, 0.121, -0.945, -0.055, -1.436, 0.360, -1.478, -1.637, -1.869]
data2 = [0.353, 3.517, 0.125, -7.545, -0.555, -1.536, 3.350, -1.578, -3.537, -1.579]
stat, p = kendalltau(data1, data2)

print(p) # Reject H0 --> Monotonic relationship

0.0031612220209069567


**Chi-Squared Test**

Tests whether two **categorical** variables are related or independent.<br>

Assumptions:<br>
- Obsrevations used in the calculation of the contingency table are independent
- 25 or more examples in each cell of the contigency table

Interpretation:<br>
- H0: Samples are independent
- H1: Samples are dependent

Python:

In [8]:
from scipy.stats import chi2_contingency

## Test dependency
table = [[10, 20, 30],[6,  9,  17]]
stat, p, dof, expected = chi2_contingency(table)

print(p) # Cannot reject H0 --> independent

0.873028283380073


#### 2.1.3 Compare Sample Means (Parametric)

Parametric statistical tests can be used to compare data samples (e.g., to proof that they are from the same population).<br>

**Student's t-test**

Tests whether the means of two independent samples are significantly different.

Assumptions:<br>
- Observations are i.i.d.
- Observations in each sample are normally distributed
- Observations in each sample have the same variance

Intepretation:<br>
- H0: Means are equal
- H1: Means are unequal

Python:

In [9]:
from scipy.stats import ttest_ind

## Test for equal means
data1 = [0.873, 2.817, 0.121, -0.945, -0.055, -1.436, 0.360, -1.478, -1.637, -1.869]
data2 = [1.142, -0.432, -0.938, -0.729, -0.846, -0.157, 0.500, 1.183, -1.075, -0.169]
stat, p = ttest_ind(data1, data2)

print(p) # Cannot reject H0 --> means are equal

0.7484698873615687


**Paired Student's t-test**

Tests whether the means of two paired samples are significantly different. Paired samples are samples where each point in one sample matches another point in the second sample. For example, in a before-after drug test, the researcher looks at patients. So, each before datapoint is related to exactly one after datapoint.<br>

Assumptions:<br>
- Observations are i.i.d.
- Observations are normally distributed
- Observations in each sample have the same variance
- Observations across each sample are paired

Interpretation:<br>
- H0: Means are equal
- H1: Means are unequal

Python:

In [10]:
from scipy.stats import ttest_rel

## Testing for equal means
data1 = [0.873, 2.817, 0.121, -0.945, -0.055, -1.436, 0.360, -1.478, -1.637, -1.869]
data2 = [1.142, -0.432, -0.938, -0.729, -0.846, -0.157, 0.500, 1.183, -1.075, -0.169]
stat, p = ttest_rel(data1, data2)

print(p) # Cannot reject H0 --> equal means

0.7459078283577478


**Analysis of Variance Test (ANOVA)**

Tests whether the means of two or more independent samples are significantly different.<br>

Assumptions:<br>
- Observations are i.i.d.
- Observations are normally distributed
- Observations in each sample have the same variance

Interpretation:<br>
- H0: Means of the samples are equal
- H1: One or more means of the samples are unequal

Python:

In [11]:
from scipy.stats import f_oneway

## Test for multiple equal means
data1 = [0.873, 2.817, 0.121, -0.945, -0.055, -1.436, 0.360, -1.478, -1.637, -1.869]
data2 = [1.142, -0.432, -0.938, -0.729, -0.846, -0.157, 0.500, 1.183, -1.075, -0.169]
data3 = [-0.208, 0.696, 0.928, -1.148, -0.213, 0.229, 0.137, 0.269, -0.870, -1.204]
stat, p = f_oneway(data1, data2, data3)

print(p) # Cannot reject H0 --> sample means are equal

0.9083957433926546


There is also a **Repeated Measures ANOVA Test**, which tests whether the means of two or more paired samples are significantly different. However, unfortunately there exists no package for this test in Python as of now.<br><br>

#### 2.1.4 Compare Sample Means (Non-Parametric)

XXX

### 2.2 Confidence Intervals

Confidence interval is another way of valuing a distance between $\bar{x}$ and $\mu$. Under $H_{0}$ the random variable $(\bar{X}-\mu)\sim Norm(0,\frac{\sigma}{\sqrt{n}})$. Thus $\displaystyle P\{|\bar{X}-\mu|<=1.96\frac{\sigma}{\sqrt{n}}\}\approx0.95$<br>

**Definition**:<br>
An interval $(\bar{x}-a_{1},\bar{x}+a_{1})$ such that $P\{\bar{X}-1.96\frac{\sigma}{\sqrt{n}}<=\mu<=\bar{X}+1.96\frac{\sigma}{\sqrt{n}}\}=1-\alpha$ is called an $(1-\alpha)$-level confidence interval for the mean.<br>

In applied machine learning, we may wish to use confidence intervals in the presentation of the skill of a predictive model. For example, a confidence interval could be used in presenting the skill of a classification model, which could be stated as:<br>
Given the sample, there is a 95% likelihood that the range x to y covers the true model accuracy.<br>
or<br>
The accuracy of the model was x +/- y at the 95% confidence level.<br>

Confidence intervals can also be used in the presentation of the error of a regression predictive model; for example:<br>
There is a 95% likelihood that the range x to y covers the true error of the model.<br>
or<br>
The error of the model was x +/- y at the 95% confidence level.<br>

These estimates of uncertainty help in two ways. First, the intervals give the consumers of the model an understanding about how good or bad the model may be. […] In this way, the confidence interval helps gauge the weight of evidence available when comparing models. The second benefit of the confidence intervals is to facilitate trade-offs between models. If the confidence intervals for two models significantly overlap, this is an indication of (statistical) equivalence between the two and might provide a reason to favor the less complex or more interpretable model.

Python:

In [1]:
from math import sqrt

# binomial confidence interval
## Calcualting interval for classification error/accuracy: interval = z * sqrt( (error * (1 - error)) / n)
interval = 1.96 * sqrt( (0.2 * (1 - 0.2)) / 50)
print('%.3f' % interval)

0.111


Therefore, the classification error of the model is 20% +/- 11% and the true classification error of the model is likely between 9% and 31%.<br>

The proportion_confint() statsmodels function an implementation of the binomial proportion confidence interval.

By default, it makes the Gaussian assumption for the Binomial distribution, although other more sophisticated variations on the calculation are supported. The function takes the count of successes (or failures), the total number of trials, and the significance level as arguments and returns the lower and upper bound of the confidence interval.

The example below demonstrates this function in a hypothetical case where a model made 88 correct predictions out of a dataset with 100 instances and we are interested in the 95% confidence interval (provided to the function as a significance of 0.05).

In [2]:
from statsmodels.stats.proportion import proportion_confint

lower, upper = proportion_confint(88, 100, 0.05)
print('lower=%.3f, upper=%.3f' % (lower, upper))

lower=0.816, upper=0.944


Lower and upper bounds of the models accuracy.<br>

**Nonparametric Confidence Interval**

The predicted variable sometimes is not normally distributed, and even when it is, the variance of the normal distribution might not be equal at all levels of the predictor value. Alternatively, we may not know the analytical way to calculate a confidence interval for a skill score.<br>

In these cases, the bootstrap resampling method can be used as a nonparametric method for calculating confidence intervals, nominally called bootstrap confidence intervals.

The bootstrap is a simulated Monte Carlo method where samples are drawn from a fixed finite dataset with replacement and a parameter is estimated on each sample. This procedure leads to a robust estimate of the true population parameter via sampling.

In this example, random values are drawn and the mean is calculated. Instead of the mean, this might just as well be a model's accuracy that is being fitted on the sample data.

In [3]:
# bootstrap confidence intervals
from numpy.random import seed
from numpy.random import rand
from numpy.random import randint
from numpy import mean
from numpy import median
from numpy import percentile

# seed the random number generator
seed(1)
# generate dataset
dataset = 0.5 + rand(1000) * 0.5
# bootstrap
scores = list()
for _ in range(100):
	# bootstrap sample
	indices = randint(0, 1000, 1000)
	sample = dataset[indices]
	# calculate and store statistic
	statistic = mean(sample)
	scores.append(statistic)
print('50th percentile (median) = %.3f' % median(scores))
# calculate 95% confidence intervals (100 - alpha)
alpha = 5.0
# calculate lower percentile (e.g. 2.5)
lower_p = alpha / 2.0
# retrieve observation at lower percentile
lower = max(0.0, percentile(scores, lower_p))
print('%.1fth percentile = %.3f' % (lower_p, lower))
# calculate upper percentile (e.g. 97.5)
upper_p = (100 - alpha) + (alpha / 2.0)
# retrieve observation at upper percentile
upper = min(1.0, percentile(scores, upper_p))
print('%.1fth percentile = %.3f' % (upper_p, upper))

50th percentile (median) = 0.750
2.5th percentile = 0.741
97.5th percentile = 0.757


In this case, the mean is likely to be between 0.741 and 0.757 (95% confidence interval).

Sources:<br>
[1] https://machinelearningmastery.com/confidence-intervals-for-machine-learning/